### PR TITLE
Add configurable OAuth 2.0 scopes

### PR DESCRIPTION
**Pull request by n2ygk**
## Hey, I just made a Pull Request!

Fixes: #4430 

- Add oauth2 config for optional scopes
- Document oauth2 config keys
- Add OAuth2 to demo app list of identity providers

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

Example screenshot using my organization's OAuth2 redirection to SAML2 login:

image

### CODE REVIEW

**Comment by n2ygk** [comment]
OK so I ran this test locally and it worked after removing callbackUrl... I'll need a little help here -- or I can just undo the config.d.ts and make the error go away.

**Comment by n2ygk** [comment]
See also #4466 regarding OIDC scopes.

**Comment by n2ygk** [comment]
Never mind. Per @rugvip https://github.com/backstage/backstage/issues/4466#issuecomment-776242919 withdrawing this PR.

**Comment by n2ygk** [comment]
Reopening this PR as I'd like to confirm with @Rugvip. These are not default scopes; but rather required scopes for a user sign-in. Specifically `openid`, `email`, and maybe `profile` are required as the login fails if no `email` claim is missing. In my use case, another ["authentication selector"](https://docs.pingidentity.com/bundle/pingfederate-93/page/vat1564002987551.html) scope is required as well to indicate which _Resource Owner_ (end user) sign-in mechanism to use.

It seems that since the backend is acting as the third-party _Client_ which implements the Authorization Code flow, storing confidential information like the clientSecret, it should also control which scopes are required for sign-in (and scope approval by) the _Resource Owner_ who delegates those scopes to the _Client_. 

Am I missing something here?

**Comment by Rugvip** [comment]
@n2ygk Yeah I think you're right that it doesn't make sense to configure the scopes required for sign-in in the frontend, so makes sense to move forward with this.

I think we should add it to the implementations and TS APIs rather than the static config though. We haven't yet made the profile transform logic customizable, but that's something else we want to do, and the scopes required to fetch the correct profile information would really be part of that. To begin with I think moving over some of the default scopes required for sign-in from the frontend to the backend would make sense. With that it's possible we might even be able to completely remove the default scopes from the frontend eventually.

For the general purpose providers like OAuth and OIDC we'd really want to be able to provide multiple instances of those with different configuration, which is a pattern we need to add to the auth-backend

**Comment by n2ygk** [comment]
@Rugvip Thanks for your guidance. To get concrete and keep things moving, is this static config PR a good start and then a future move toward the methods you describe (which I don't understand:-)? It at least allows me to enable OAuth2/OIDC login.  Or, specific pointers to where these changes should happen instead would be appreciated and I can take a crack at it.

Also, a pointer on why the config.d.ts tests are failing would be appreciated since they pass locally. Or I can just undo that change and go back to using the generic `[key: string]: string`.

File: app-config.yaml
Comment (Rugvip) [inline app-config.yaml:309]: No need to put this in an env var, could just add a commented out example value here instead like in the changeset

**Review by Rugvip — APPROVED** [review]
Yeah don't have any other good answer for how to move forward with this to unblock you right now tbh, but it will be broken in the future : /

Regarding the config schema it's awkwardness of how we define the default config in the main repo. I pushed a tweak to the workflows which should hopefully fix it

**Review by Rugvip — APPROVED** [review]
:shipit:

**Pull request merged**
